Fix 867 Added error messages when add rule fails

juanpablosalas opened the pull request:
When add rule fails due to permission being denied, an error message detailing the exception is displayed using the PermissionResult described in [rucio/rucio#65680](https://github.com/rucio/rucio/issues/6580).

Relevant issues: #867 #545 

Fixes #545

ericvaandering commented on src/policy/CMSRucioPolicy/permission.py:
```diff
@@ -187,7 +188,7 @@ def _get_rule_size(rules):
     # prevent rule creation under 'User AutoApprove' for rules without ask_approval
     if not kwargs["ask_approval"]:
-        return False
+        return PermissionResult(False,'ask_approval argument is necessary under User AutoApprove.')
```
Put spaces after commas in python

ericvaandering replied:
Everywhere else too

ericvaandering commented on src/policy/CMSRucioPolicy/permission.py:
```diff
@@ -267,7 +268,9 @@ def _get_rule_size(rules):
                     single_rse_rule_size_threshold)
-                return False
+                return PermissionResult(False,
+                        'Single RSE usage exceeded for auto approve rules for account %s and RSE %s, this_rse_autoapprove_usage, size_of_rule, single_rse_rule_size_threshold: %s, %s, %s',
```
Use an f-string to improve readability. Keep the lines shorter.

ericvaandering replied:
Same elsewhere too

ericvaandering commented on src/policy/CMSRucioPolicy/permission.py:
```diff
@@ -341,7 +349,7 @@ def perm_add_rule(issuer, kwargs, *, session: "Optional[Session]" = None):
 
     if any_tape and kwargs['lifetime'] is not None:
-        return False
+        return PermissionResult(False,"A rule cannot be created for a tape RSE with a lifetime.")
```
Why not? And do we want users putting data on tape anyhow?

juanpablosalas replied:
I changed the error message to convey this is because of DM policy. Users with AutoApprove cannot create tape rules but this line should apply to users with permissions to create tape rules (i.e. transfer_ops).

ericvaandering commented on src/policy/CMSRucioPolicy/permission.py:
```diff
@@ -263,21 +264,16 @@ def _get_rule_size(rules):
-                    account, rse_expression, this_rse_autoapprove_usage, size_of_rule,
-                    single_rse_rule_size_threshold)
-                return False
+                logging.warning(f'Single RSE usage exceeded for auto approve rules for account {account} and RSE {rse_expression}, this_rse_autoapprove_usage, size_of_rule, single_rse_rule_size_threshold: {this_rse_autoapprove_usage}, {size_of_rule}, {single_rse_rule_size_threshold}')
```
These lines are too long. Please look at https://stackoverflow.com/questions/10660435/how-do-i-split-the-definition-of-a-long-string-over-multiple-lines and keep the lines to 100 characters or so.

I've prepended the f-strings in the example below with f, but note you still have to put spaces in the strings in the right places.

``` 
s = (f"this is a very"
      f"long string too"
      f"for sure ..."
     )
```

ericvaandering replied:
Also, just do you know. logging with f-strings or anything other than the form you are replacing is frowned on. The reason is that only if the message is going to be printed does the string get constructed. Now in this case, In my opinion, the readability of the code is more important since we know this message is going to be output infrequently. But if you were doing something like a DEBUG message for every transfer, then you would not want to be constructing a bunch of strings no one was ever going to see.

juanpablosalas replied:
Thank you, Eric. Do you suggest saving the string as a variable and then input that variable in the logging and the PermissionResult so as not to repeat code?

ericvaandering replied:
No. (My second comment was just a little about python an best practices, nothing for this PR)

ericvaandering commented on src/policy/CMSRucioPolicy/permission.py:
```diff
@@ -313,11 +307,12 @@ def perm_add_rule(issuer, kwargs, *, session: "Optional[Session]" = None):
             if rse_attr.get('requires_approval', False):
-                return False
+
+                return PermissionResult(False, "One or more of the RSEs matching the expression needs approval and thus the rule cannot be created (i.e. ask_approval is missing).")
```
Even this one should be split in two.

dynamic-entropy commented on helm/rucio-consistency/Chart.yaml:
```diff
@@ -15,7 +15,7 @@ type: application
 # to the chart and its templates, including the app version.
 # Versions are expected to follow Semantic Versioning (https://semver.org/)
-version: 3.4.0
+version: 3.4.1
```
These changes are unrelated to the PR. Should not be part of the same PR.

dynamic-entropy commented on helm/rucio-consistency/templates/consistency-configmap.yaml:
```diff
@@ -83,3 +83,4 @@ data:
 {{- end }}
 {{- end }}
 {{- end }}
+{{- end }}
```
new line? Also was this helm chart ever broken?

dynamic-entropy commented on src/policy/CMSRucioPolicy/permission.py:
```diff
@@ -232,17 +233,17 @@ def _get_rule_size(rules):
     # Check if the account is banned
     if has_account_attribute(account, 'auto_approve_banned', session=session):
-        return False
+        return PermissionResult(False, "The account is banned.")
```
I would add instructions for further action - such as creating a ticket with DM in jira.

dynamic-entropy commented on src/policy/CMSRucioPolicy/permission.py:
```diff
@@ -232,17 +233,17 @@ def _get_rule_size(rules):
 
     if kwargs['lifetime'] is None:
-        return False
+        return PermissionResult(False, "lifetime argument is None.")
```
I think this should be worded in passive voice with an explanation:
Something on the lines of "Rules with infinite (without) lifetime cannot be created under so and so activity".

dynamic-entropy replied:
same thing applies to other messages in general.

dynamic-entropy commented on src/policy/CMSRucioPolicy/permission.py:
```diff
@@ -232,17 +233,17 @@ def _get_rule_size(rules):
 
     if kwargs['lifetime'] > rule_lifetime_threshold:
-        return False
+        return PermissionResult(False, "The rule lifetime exceeds the treshold.")
```
Since these messages are for users, robots do not use the "user autoapproce" activity (I hope, haha).
We should point them to the docs that explain and mention the thresholds (or you can print the limit here too if you want).

dynamic-entropy replied:
Same this applies to other messages as well.

dynamic-entropy reviewed (commented):
Left a couple of comments. 
I believe the unrelated commits for the helm charts should be done in a separate PR / removed if not intentional.

Cheers

juanpablosalas commented:
Thanks for the comments, I changed the messages to be a little more self explanatory for users. Also, the helm chart changes are definitely not a part of this PR, I think they got mixed up with a commit Eric did but I fixed it now in this PR.

ericvaandering commented:
I had to back this out of the release. It was not importing cleanly 

```
[Mon Apr 14 19:42:39.992312 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616] Traceback (most recent call last):                                
[Mon Apr 14 19:42:39.992340 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]   File "/usr/local/lib/python3.9/site-packages/rucio/core/permissi
on/__init__.py", line 74, in <module>                                                                                                                         
[Mon Apr 14 19:42:39.992344 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]     module = importlib.import_module(policy)                      
[Mon Apr 14 19:42:39.992347 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]   File "/usr/lib64/python3.9/importlib/__init__.py", line 127, in 
import_module                                                                                                                                                 
[Mon Apr 14 19:42:39.992349 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]     return _bootstrap._gcd_import(name[level:], package, level)   
[Mon Apr 14 19:42:39.992352 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]   File "<frozen importlib._bootstrap>", line 1030, in _gcd_import 
[Mon Apr 14 19:42:39.992356 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]   File "<frozen importlib._bootstrap>", line 1007, in _find_and_lo
ad                                                                                                                                                            
[Mon Apr 14 19:42:39.992359 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]   File "<frozen importlib._bootstrap>", line 986, in _find_and_loa
d_unlocked                                                                                                                                                    
[Mon Apr 14 19:42:39.992368 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]   File "<frozen importlib._bootstrap>", line 680, in _load_unlocke
d                                                                                                                                                             
[Mon Apr 14 19:42:39.992371 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]   File "<frozen importlib._bootstrap_external>", line 850, in exec
_module                                                                                                                                                       
[Mon Apr 14 19:42:39.992374 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]   File "<frozen importlib._bootstrap>", line 228, in _call_with_fr
ames_removed                                                                                                                                                  
[Mon Apr 14 19:42:39.992377 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]   File "/opt/rucio/policy/CMSRucioPolicy/permission.py", line 33, 
in <module>                                                                                                                                                   
[Mon Apr 14 19:42:39.992380 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]     from rucio.core.permission import PermissionResult            
[Mon Apr 14 19:42:39.992389 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616] ImportError: cannot import name 'PermissionResult' from partially 
initialized module 'rucio.core.permission' (most likely due to a circular import) (/usr/local/lib/python3.9/site-packages/rucio/core/permission/__init__.py)
[Mon Apr 14 19:42:39.992393 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]                                                                   
[Mon Apr 14 19:42:39.992396 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616] During handling of the above exception, another exception occurred
:                                                                                                                                                             
[Mon Apr 14 19:42:39.992398 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]                                                                   
[Mon Apr 14 19:42:39.992401 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616] Traceback (most recent call last):                                
[Mon Apr 14 19:42:39.992413 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]   File "/usr/local/lib/python3.9/site-packages/rucio/web/rest/ping
.py", line 18, in <module>                                                                                                                                    
[Mon Apr 14 19:42:39.992415 2025] [wsgi:error] [pid 56:tid 619] [remote 10.100.2.157:48616]     from rucio.web.rest.flaskapi.v1.ping import blueprint as metri
```
